hivemq-edge/src/main/java/com/hivemq/edge/modules/adapters/impl/polling/PollingTask.java

```diff
@@ -0,0 +1,210 @@
+        if (!stopBecauseOfTooManyErrors) {
+            reschedule(errorCountTotal);
+        }
+    }
+
+
+    private void handleExceptionDuringPolling(final @NotNull Throwable throwable) {
+        int errorCountTotal = applicationErrorCount.incrementAndGet();
+        final int maxErrorsBeforeRemoval = sampler.getMaxErrorsBeforeRemoval();
+        // case 1: Unlimited retry (maxErrorsBeforeRemoval < 0) or less errors than the limit
+        if (maxErrorsBeforeRemoval < 0 || errorCountTotal <= maxErrorsBeforeRemoval) {
+            if (log.isDebugEnabled()) {
+                log.debug("Application Error {} in sampler {} -> {}",
+                        errorCountTotal,
+                        sampler.getAdapterId(),
+                        throwable.getMessage(),
+                        throwable);
+            }
+            reschedule(errorCountTotal);
+            notifyOnError(sampler, throwable, true);
+        } else {
+            log.info(
+                    "Detected '{}' recent errors when sampling. This exceeds the configured limit of '{}'. Sampling for adapter with id '{}' gets stopped.",
+                    errorCountTotal,
+                    maxErrorsBeforeRemoval,
+                    sampler.getAdapterId());
+            notifyOnError(sampler, throwable, false);
+            // no rescheduling
+        }
+
+    }
+
+    private void notifyOnError(
+            final @NotNull ProtocolAdapterPollingSampler sampler, final @NotNull Throwable t, boolean continuing) {
+        try {
+            sampler.error(t, continuing);
+        } catch (Throwable samplerError) {
+            if (log.isInfoEnabled()) {
+                log.info("Sampler Encountered Error In Notification", samplerError);
+            }
+        }
+    }
+
+    private void reschedule(int errorCountTotal) {
+        final long delayInMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - nanosOfLastPolling);
+        // a negative delay means that the last polling attempt took longer to be processed than the specified delay between polls
+        if (delayInMillis < 0) {
+            log.warn(
+                    "Polling for protocol adapter '{}' can not keep up with the specified '{}' interval, because the polling takes too long.",
+                    sampler.getAdapterId(),
+                    sampler.getPeriod());
+            eventService.createAdapterEvent(sampler.getAdapterId(), sampler.getProtocolId())
+                    .withMessage(String.format(
+                            "Polling for protocol adapter '%s' can not keep up with the specified '%d' ms interval, because the polling takes too long.",
+                            sampler.getAdapterId(),
+                            sampler.getPeriod()))
+                    .withSeverity(Event.SEVERITY.WARN)
+                    .fire();
+        }
+
+        long nonNegativeDelay = Math.max(0, delayInMillis);
+
+        if (errorCountTotal == 0) {
+            schedule(nonNegativeDelay);
```

Comment:
Don't you want to consider `sampler.getPeriod()` here? Otherwise, if the polling returns quickly enough, there would be barely any delay between polls.
Looks like `ProtocolAdapterPollingSampler#getPeriod()` is not respected here in general, seems accidental to me?

Reply:
the delay should include the period time already. Likely I made a mistake here.